README.md
# AI Code Bot 🤖

## Your AI-powered coding sidekick

AICodeBot is a coding assistant designed to make your coding life easier. Think of it as your AI version of a pair programmer. Perform code reviews, create helpful commit messages, debug problems, and help you think through building new features. A team member that accelerates the pace of development and helps you write better code.

We've planned to build out multiple different interfaces for interacting with AICodeBot. To start, it's a [command-line tool](https://pypi.org/project/aicodebot/) that you can install and run in your terminal, and a [GitHub Action for Code Reviews](https://github.com/marketplace/actions/aicodebot-code-review).

Status: This project is in its early stages, but it already improves the software development workflow, and has a healthy roadmap of features (below).

We're using AICodeBot to build AICodeBot, and it's upward spiraling all the time.️ We're looking for contributors to help us build it out. See [CONTRIBUTING](https://github.com/TechNickAI/AICodeBot/blob/main/CONTRIBUTING.md) for more.

## Current features - how you can use it today

| Task | Status |
| --- | --- |
| Generating quality commit messages | ✅ |
| Thinking through tasks as a pair programmer | ✅ |
| Coding with a small number of files | ✅ |
| Debugging | ✅ |
| Doing code reviews | ✅ |
| Explaining code | ✅ |
| Writing tests | ✅ |
| Integration with GitHub | ✅ |
| Modifying local files | Early stages |
| Searching the internet for answers | In Progress |
| Reading library documentation | Early stages |
| Coding with a large number of files |  As LMs get larger token limits |
| Writing senior developer level code | Eventually |
| Major refactors | Eventually |
| Build entire apps | Eventually |
| Replace Developers | Nope |

### AI Sidekick 🦸‍♂️

 `aicodebot sidekick` Your AI-powered coding assistant. It's designed to help you with your coding tasks by providing context-aware suggestions and solutions. Think ChatGPT with the ability to read the local repository for context.

 By default it will pass along a directory of files from the current working directory, and you can also pass in a list of files to use as context for the task you are working on. For example:

 ```bash
 aicodebot sidekick file1.py file2.py
 ```

In this example, the sidekick will read the contents of file1.py and file2.py and use them to provide context-aware answers.

Pro-tips:

* Add your README.md to the list of files to get context-aware answers.
* Paste in error messages from log files or stack traces and it will help you debug
* Brainstorm new features
* Have it write unit tests for you that cover all the cases

This feature is in its early phases right now, but it's already useful. We'll be adding support for tools that the sidekick can use, including GitHub integration, ingesting repository specific domain knowledge, writing local files, and more. For now, it just *reads* files and provides suggestions.

### AI-Assisted Git Commit 📝

`aicodebot commit` improves the git commit process. It will run pre-commit for you to check syntax, and then generate a commit message for you based on the changes you've made. In about as much effort as typing "fix bug" for the commit message, you will get a high-quality commit message that thoroughly describes the change.

### AI-Assisted Code Review 👀

`aicodebot review` will run a code review on your code and suggest improvements. By default, it will look at [un]staged changes, and you can also supply a specific commit hash to review.
Its goal is to suggest how to make the code better, and we've found that it often teaches us new things about our code and makes us better programmers. It is a great way to get a second set of robot eyes on your code.

### AI-Assisted Debugging 🐞

`aicodebot debug $command` will run the $command and capture the log output. It will pass the error message, stack trace, command output, etc. over to the AI and respond with some suggestions on how to fix it.

[![PyPI version](https://badge.fury.io/py/aicodebot.svg?0.6.2)](https://badge.fury.io/py/aicodebot)

## Installation and Usage

To install AICodeBot, run:

`pip install aicodebot`

And then, run `aicodebot configure` to get started.

```bash
Usage: aicodebot [OPTIONS] COMMAND [ARGS]...

Options:
  -V, --version  Show the version and exit.
  -h, --help     Show this message and exit.
  -d, --debug    Enable langchain debug output

Commands:
  alignment       A message from AICodeBot about AI Alignment ❤ + 🤖.
  commit          Generate a commit message based on your changes.
  configure       Create or update the configuration file
  debug           Run a command and debug the output.
  review          Do a code review, with [un]staged changes, or a...
  sidekick        Coding help from your AI sidekick
```

### Open AI key setup

The first time you run it, you'll be prompted to enter your OpenAI API Key, which is required, as we use OpenAI's large language models for the AI. You can get one for free by visiting your [API key settings page](https://platform.openai.com/account/api-keys).

Note: You will be billed by OpenAI based on how much you use it. Typical developers will use less than $10/month - which if you are a professional developer you'll likely more than make up for with saved time and higher quality work. See [OpenAI's pricing page](https://openai.com/pricing/) for more details. Also, see the note about increasing your token size and using better language models below.

## Integration with GitHub Actions

How about automated code reviews on every commit? You can have AICodeBot run as a GitHub action on your repository. See [The AICodeBot GitHub Action for Code Reviews](https://github.com/marketplace/actions/aicodebot-code-review). It will look at every commit and pull request, and then either:

* ✅ Pass
* 🗒️ Comments - pass the action, but leave a comment with minor issues, suggestions, or improvements.
* ❌ Fail the action - if a serious issue is found, it will leave a comment and let you know about something that should be addressed.

## Roadmap of Upcoming Features ️

### Code Workflow Improvements

* [X] **Assisted Git Commit**: Automatically generate a commit message based on the changes you've made
* [X] **Assisted Debugging**: Run a command with aicodebot and it captures the log message and tries to figure out what's going on from the error message.  Eventually, it could also suggest fixes for the error and make the changes for you. Try it out with `aicodebot debug $command`
* [X] **Code Review**: Provides feedback on potential issues in code,  and suggests improvements to make it better.
* [ ] **Dependency Management**: Updating dependencies to their latest versions with pull requests that run tests.
* [ ] **Documentation Generation**: Generates comprehensive documentation for code, including docstrings, README files, and wiki pages.
* [ ] **Performance Optimization Suggestions**: Suggests potential performance optimizations for code.
* [X] **Test Generation**: Generates unit tests for code, improve test coverage.
* [X] **Integration with CI/CD pipelines**: Integrates with CI/CD pipelines to automate tasks like code review, testing, and deployment (via GitHub Actions). Eventually: Fix the build automatically when there are small errors.
* [X] **Rubber Ducky Chat Bot**: Helps developers think through design issues by providing a conversational interface to discuss and solve problems, using data from the current repository.
* [X] **Linting/Formatting**: Checks code for linting errors and automatically fixes them where possible (via pre-commit)
* [ ] **Automatically Generate ChangeLogs and Release Notes**: Generates release notes and changelogs based on commit messages and code changes.

### User Interfaces

* [X] **Command-line, installable via pip**: aicodebot can be installed as a Python package using `pip install aicodebot`
* [ ] **Mention the @aicodebot GitHub user**: Mentioning the [@aicodebot](https://pypi.org/project/aicodebot/) GitHub user in a comment will trigger it to perform a task, review code, or pull in an appropriate GIF.
* [X] **Callable as a GitHub action**: Can be called as a GitHub action to perform tasks on GitHub repositories. [AICodeBot Action](https://github.com/marketplace/actions/aicodebot-code-review)
* [ ] **Jupyter Notebook Extension**: Provides a Jupyter Notebook extension that can be used to debug code in the notebook.
* [X] **Chat**: CLI chat interface that knows the context of your codebase and can answer questions about it. No more going back and forth between ChatGPT and command-line.
* [ ] **Slack Bot**: Interacts with aicodebot via slack, sends notifications, performs tasks, and provides real-time assistance to developers.
* [ ] **Bug Report service integrations**: Listen for bug reports from Sentry, [Honeybadger](http://honeybadger.io), and other bug reporting tools and automatically create issues, assign them to developers, and notify them via Slack. Eventually: FIX the bug automatically and notify the team.

### Repository / Project Management

* [ ] **Project best practices**: Suggest things like pre-commit, linting, license, CI/CD, etc. Eventually: Implement them for you.
* [ ] **Manage GitHub Issues**: Provides Level 1 triage of incoming issues on GitHub, including tagging, assigning, and responding with FAQs. It could also escalate issues to human developers when necessary, and provide nudges for tasks that need more input.
* [ ] **Welcome new contributors**: Automatically welcome new contributors to the project, find out what they're interested in, and suggest issues for them to work on.

### Fun

* [X] **Alignment**: Gives a heart-centered inspirational message about how we can build AI in a way that aligns with humanity. Try it out with `aicodebot alignment`.
* [ ] **Telling Jokes**: We've gotta figure out how to teach language models about humor. :)
* [ ] **Supportive Encouragement**: High fives and kudos for a job well done
* [ ] **GIF Reactions**: Reacts to messages with relevant and fun gifs.

<img src="https://camo.githubusercontent.com/6fc1e79b4aa226b24a756c4c8e20e5b049301a930449a7321d3e45f516e61601/68747470733a2f2f74656e6f722e636f6d2f766965772f6b746f2d6b6f756e6f746f72692d6b6f756e6f746f7269746f6b656e2d6c626f772d73746f726b686f6c646572732d6769662d32353637363438332e676966" width="25%">

## Alignment ❤️ + 🤖

Technology itself is amoral; it just imbues the values of the people who create it. We believe that AI should be built-in a way that aligns with humanity, and we're building AICodeBot to help us do just that. We're building from a heart-centered space, and contributing to the healthy intersection of AI and humanity.

### What it's NOT

`aicodebot` is a tool for developers, not a replacement for them. It's not going to replace your job, but it will make your job easier and more fun. It won't take over the world, but it will help us build a better one. See the *Alignment* section below for more.

⚠️ AICodeBot currently uses OpenAI's ChatGPT large language models, which can hallucinate and be confidently wrong. Sometimes AICodeBot does dumb things, so it's mostly *reading* and *advising* and not yet *writing*. Much like Tesla's "Full Self Driving", you must keep your hands on the wheel.

It's also not a "build a site for me in 5 minutes" tool that takes a well-constructed prompt and builds a scaffold for you. There are [other tools](https://github.com/AntonOsika/gpt-engineer) for that. It's not a no-code platform. Instead, AICodeBot is built to work with existing codebases and the git-commit level. It's designed to multiply the effectiveness of capable engineers.

## Configuring the language model to use

Not all OpenAI accounts have GPT-4 API access enabled. By default, AICodeBot will use GPT-4. If your OpenAI account supports it, we will check the first time you run it. If your OpenAI API does not support GPT-4, you can ask to be added to the waitlist [here](https://openai.com/waitlist/gpt-4-api). In our testing, GPT-4 is the best model and provides the best results.

To specify a different model, you can set the `language_model` in your `$HOME/.aicodebot.yaml` file. For example:

```yaml
openai_api_key: sk-*****
language_model: gpt-3.5-turbo
personality: Stewie
version: 1.2
```

You can also use openrouter.ai to get access to advanced models like GPT-4 32k and Anthropic's 100k model for larger context windows. See [openrouter.ai](https://openrouter.ai) for more details. Here's a sample config:

```yaml
openai_api_key: sk-*****
openrouter_api_key: sk-or-****
language_model_provider: OpenRouter
language_model: openai/gpt-4-32k # or anthropic/claude-2 for 100k token limit
personality: Stewie
version: 1.2
```

Note: We'll be adding more options for AI models in the future, including those that can be run locally, such as [GPT4all](https://gpt4all.io/) and HuggingFace's [Transformers](https://huggingface.co/transformers/).

### Understanding Tokens and Using Commands Efficiently

In AI models like OpenAI's GPT-4, a "token" is a piece of text, as short as a character or as long as a word. The total tokens in an API call, including input and output, affect the cost, time, and whether the call works based on the maximum limit.

Each model has a maximum token limit. For example, GPT-3.5 has a limit of 4096 tokens, and GPT-4 has a token limit of 8192 tokens. If a conversation exceeds this limit, you must reduce your text until it fits.

When using commands like the Sidekick command in AICodeBot, which allows you to pass in files for context, it's important to manage your tokens effectively. Due to token limits, it's not feasible to load your entire codebase. Instead, you should only load the specific files that are relevant to the task you're working on.
This ensures that the AI model can process your request efficiently and provide the most relevant suggestions for your current task.

### How can I get a larger token limit?

Do you need a larger context window for your task? Are you running into token limits and getting a message like this?

```bash
The context is too large (21414) for any of the models supported by your API key. 😞
```

There are a couple of things you can do:

1. Load fewer files into the context (only what you need to work with)
2. Apply for GPT-4-32k access from OpenAI by contacting them.
3. Use openrouter.ai - this allows you to use the full power of GPT-4-32k, which offers a 4x larger context window. See [openrouter.ai](https://openrouter.ai) for more details. Once you sign up and set your `openrouter_api_key` in your `$HOME/.aicodebot.yaml` file, you can have access to larger models. Soon we will have support for Claude 2, which has a 100k token limit.

## Development / Contributing

We'd love your help! If you're interested in contributing, here's how to get started. See [CONTRIBUTING](https://github.com/TechNickAI/AICodeBot/blob/main/CONTRIBUTING.md) for more details.

## Docker

Assumes you have changes in current working dir that are already added.

```shell
docker build -t aicodebot .
docker run -v ~/.aicodebot.yaml:/home/user/.aicodebot.yaml -v .:/app aicodebot commit -y
```


CHANGELOG.md
# CHANGELOG



## v0.15.0 (2023-07-23)

### Unknown

* Bump version to 0.15.0 ([`313be38`](https://github.com/TechNickAI/AICodeBot/commit/313be38a3881500f2ae54572a26f5acf4728ac2f))

* Update ruff in pre-commit ([`f75fa25`](https://github.com/TechNickAI/AICodeBot/commit/f75fa2518378b27dff8a20a1a63f0043efbd6c35))

* Black formatting matters ([`151a6cf`](https://github.com/TechNickAI/AICodeBot/commit/151a6cf07618dc449e00af2d4da98ebb1fadfa50))

* Let&#39;s have pre-commit use the black configuration files instead of duplicating the line length here ([`ca66bac`](https://github.com/TechNickAI/AICodeBot/commit/ca66bac1bbe309ea4e28769fe768be232f5c9175))

* Add /command interface to sidekick with tools for adding/removeing files

In `aicodebot/helpers.py`, a new class `SidekickCompleter` has been added to provide command completion functionality in the sidekick feature.

In `aicodebot/learn.py`, the error messages have been made more informative and user-friendly.

The `requirements.in` and `requirements.txt` files have been updated with the addition of the `humanize` library. ([`f596f5e`](https://github.com/TechNickAI/AICodeBot/commit/f596f5e75d9298ca87d2a324dba47b100f162836))

* Adjust line-length in black and isort configurations 📏

The line-length in both the black and isort configurations in the pyproject.toml file has been reduced from 120 to 115. This change aims to improve code readability and maintainability. 📚 ([`28e056e`](https://github.com/TechNickAI/AICodeBot/commit/28e056e6affe8f2a92c0fc1113cae0d50c7ff500))

* Merge pull request #51 from TechNickAI/dependabot/pip/langchain-0.0.238

Bump langchain from 0.0.231 to 0.0.238 ([`7cd36d0`](https://github.com/TechNickAI/AICodeBot/commit/7cd36d06f68ef169902c2bf4adf39f2a4e00b566))

* Update pytest command to record new API responses 📝

Modified the pytest command in the GitHub workflow to include the `--record-mode=new_episodes` option. This change allows pytest to record new API responses, accommodating for instances where the OpenAI API calls a different host. This should improve the accuracy of our tests and coverage reports. ([`ba90ae3`](https://github.com/TechNickAI/AICodeBot/commit/ba90ae3543f3e330948e1f1d00e13d0bd4618464))

* Refine docstrings and welcome message in cli.py 📝

Updated the docstrings for the &#39;alignment&#39; and &#39;configure&#39; functions to better reflect their purpose. Also, enhanced the welcome message in the &#39;setup_config&#39; function for a more personalized user experience. ([`a576ed7`](https://github.com/TechNickAI/AICodeBot/commit/a576ed740624fa9aef6dc3dd99c82507568b2770))

* Proper indentation so test_parse_github_url is run ([`7a4e96c`](https://github.com/TechNickAI/AICodeBot/commit/7a4e96c5950b47cf734782396fcb11b3b57a5889))

* Update tests to use VCR and pytest-recording for API mocking

Resolves #11

The tests have been updated to use the VCR package for recording and replaying API responses during testing.

This allows the tests to run without hitting the real OpenAI API, making them faster and not consuming API quota. The responses are cached in yaml files that can be committed to version control.

To enable this:

- Added a pytest fixture to configure VCR
- Updated the test decorators from `@pytest.mark.skipif` to `@pytest.mark.vcr()`
- Recorded new cassettes for each test

Now the tests will replay cached responses instead of making live calls.

This is better for CI and local testing.
👍 ([`2547051`](https://github.com/TechNickAI/AICodeBot/commit/2547051e2908942ae6c9fbddf2969215d7a54697))

* Update &#39;Her&#39; personality description in prompts.py

Enhanced the &#39;Her&#39; personality description in the prompts.py file to include more nuanced characteristics. The AI now embodies a playful, witty, and sultry persona, while maintaining professional boundaries. This change adds depth to the AI&#39;s interaction, making it more engaging and relatable. 🎭👩‍💼🌟 ([`e4ad783`](https://github.com/TechNickAI/AICodeBot/commit/e4ad783ee88814517bc30a9eb6c042613c00d481))

* Increase the default token size for sidekick, to get better answers. ([`9fe8e8b`](https://github.com/TechNickAI/AICodeBot/commit/9fe8e8b877323cea149ee468505a51b4f3f84250))

* Bump langchain from 0.0.231 to 0.0.238

Bumps [langchain](https://github.com/hwchase17/langchain) from 0.0.231 to 0.0.238.
- [Release notes](https://github.com/hwchase17/langchain/releases)
- [Commits](https://github.com/hwchase17/langchain/compare/v0.0.231...v0.0.238)

---
updated-dependencies:
- dependency-name: langchain
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] &lt;support@github.com&gt; ([`2e61436`](https://github.com/TechNickAI/AICodeBot/commit/2e61436e6e86d3643692e9dd3b7724d4e3fd61e4))

* Merge pull request #48 from TechNickAI/dependabot/pip/click-8.1.6

Bump click from 8.1.4 to 8.1.6 ([`7fd0408`](https://github.com/TechNickAI/AICodeBot/commit/7fd0408ae4b7a04118f839ee4099c9691e07fb6a))


## v0.14.1 (2023-07-21)

### Unknown

* Bump version to 0.14.1 ([`75cb9a4`](https://github.com/TechNickAI/AICodeBot/commit/75cb9a4b0584ffeeb01d8ecfb94e03ffda0b0989))

* Skip Python 3.9 in GitHub Actions build matrix 🏗️

Due to a build failure with Python 3.9 in GitHub Actions, we&#39;re temporarily skipping it. The build matrix now only includes Python 3.10 and 3.11. We&#39;ll revisit this once the issue with Python 3.9 is resolved. ([`529f07b`](https://github.com/TechNickAI/AICodeBot/commit/529f07b247087a701e9369fc80ad951053ec6d32))

* Try turning off cache to see if that fixes python 3.9 build issue ([`9a4e45d`](https://github.com/TechNickAI/AICodeBot/commit/9a4e45d276c724d1421e82cdf62115dc8b46cb80))

* Remove the check for main in setup.py, we use aicodebot/__init__.py to store the version now ([`db8cde8`](https://github.com/TechNickAI/AICodeBot/commit/db8cde8a45462b3d9bc7f6d93f92143cfe8e94c2))

* Update Python version support and cleanup setup.py 🧹


aicodebot/lm.py
from aicodebot import AICODEBOT_NO_EMOJI
from aicodebot.config import read_config
from aicodebot.helpers import logger
from langchain import HuggingFaceHub
from langchain.chains import LLMChain
from langchain.chat_models import ChatOpenAI
from langchain.chat_models import ChatLiteLLM
from langchain.memory import ConversationTokenBufferMemory
from openai.api_resources import engine
import functools, openai, os, tiktoken

DEFAULT_RESPONSE_TOKENS = 512
DEFAULT_MEMORY_TOKENS = DEFAULT_RESPONSE_TOKENS * 2
DEFAULT_CONTEXT_TOKENS = DEFAULT_RESPONSE_TOKENS * 4
PRECISE_TEMPERATURE = 0.05
CREATIVE_TEMPERATURE = 0.6


class LanguageModelManager:
    """A class for interacting with language models."""

    provider = model_name = _memory = None

    OPENAI = "OpenAI"
    OPENROUTER = "OpenRouter"
    HUGGINGFACE_HUB = "HuggingFace Hub"
    AZURE = "Azure"
    PALM = "Palm"
    BEDROCK = "Bedrock"
    ANTHROPIC = "Anthropic"
    AI21 = "A21"
    NLP_CLOUD = "NLP Cloud"
    REPLICATE = "Replicate"
    COHERE = "Cohere"
    TOGETHER_AI = "Together AI"
    ALEPH_ALPHA = "Aleph Alpha"
    
    PROVIDERS = [OPENAI, OPENROUTER, HUGGINGFACE_HUB, AZURE, PALM, BEDROCK, ANTHROPIC, AI21, NLP_CLOUD, REPLICATE, COHERE, TOGETHER_AI, ALEPH_ALPHA]
    DEFAULT_MODEL = "gpt-4"
    DEFAULT_PROVIDER = OPENAI

    def __init__(self, model_name=None, provider=None):
        self.model_name = model_name
        self.provider = provider
        if not self.model_name:
            self.read_model_config()

    # --------------------------------- Factories -------------------------------- #

    def chain_factory(
        self,
        prompt,
        response_token_size=None,
        temperature=PRECISE_TEMPERATURE,
        streaming=False,
        callbacks=None,
        chat_history=False,
    ):
        language_model = self.model_factory(response_token_size, temperature, streaming, callbacks)
        if chat_history:
            memory = self.get_memory(language_model)
        else:
            memory = None
        return LLMChain(llm=language_model, prompt=prompt, memory=memory)

    def model_factory(
        self,
        response_token_size=None,
        temperature=PRECISE_TEMPERATURE,
        streaming=False,
        callbacks=None,
    ):
        """Get a model object for the specified model name."""

        # We support multiple approaches for using language models:
        # 1. OpenAI API
        # 2. Open Router API
        # 3. SOON Hugging Face API
        # 4. FUTURE - Local models

        provider, model_name = self.read_model_config()

        try: 
            return get_litellm_model(
                model_name,
                response_token_size=response_token_size,
                temperature=temperature,
                streaming=streaming,
                callbacks=callbacks,
            )
        else:  # pragma: no cover
            raise ValueError(f"Provider {provider} is not one of: {self.PROVIDERS}")

    def get_huggingface_hub_model(
        self,
        model_name,
        response_token_size=None,
        temperature=PRECISE_TEMPERATURE,
        streaming=False,
        callbacks=None,
    ):
        # Support for HuggingFace Hub. It works. Barely. Don't expect good results.
        # And there is no streaming.
        # If you want to play with it, add the following to your ~/.aicodebot.yaml file:
        """
        language_model_provider: HuggingFace Hub
        huggingface_api_key: hf_xxxxx
        # This is about the only model I could get to work
        language_model: google/flan-t5-xxl
        """

        api_key = self.get_api_key("huggingface_api_key")
        return HuggingFaceHub(
            huggingfacehub_api_token=api_key,
            repo_id=model_name,
            model_kwargs={
                "temperature": temperature,
                "max_length": response_token_size,
                "max_new_tokens": response_token_size,
            },
        )

    def get_memory(self, llm, token_limit=DEFAULT_MEMORY_TOKENS, memory_key="chat_history", input_key="task"):
        """Initializes a memory object with the specified parameters."""
        if not self._memory:  # Re-use the same memory object
            self._memory = ConversationTokenBufferMemory(
                memory_key=memory_key,
                input_key=input_key,
                llm=llm,
                max_token_limit=token_limit,
                human_prefix="Software Engineer",
                ai_prefix=AICODEBOT_NO_EMOJI,
            )
        return self._memory
    
    def get_litellm_model(
        self,
        model_name,
        response_token_size=None,
        temperature=PRECISE_TEMPERATURE,
        streaming=False,
        callbacks=None,
    ):
        """Get a LiteLLM model object for the specified model name."""
        return ChatLiteLLM(
            model=model_name,
            max_tokens=response_token_size,
            temperature=temperature,
            streaming=streaming,
            callbacks=callbacks,
        )

    def get_openai_model(
        self,
        model_name,
        response_token_size=None,
        temperature=PRECISE_TEMPERATURE,
        streaming=False,
        callbacks=None,
    ):
        """Get an OpenAI model object for the specified model name."""
        api_key = self.get_api_key("OPENAI_API_KEY")

        return ChatOpenAI(
            openai_api_key=api_key,
            model=model_name,
            max_tokens=response_token_size,
            temperature=temperature,
            streaming=streaming,
            callbacks=callbacks,
        )

    def get_api_key(self, key_name):
        # Read the api key from either the environment or the config file
        key_name_upper = key_name.upper()
        api_key = os.getenv(key_name_upper)
        if api_key:
            return api_key
        else:
            config = read_config() or {}
            key_name_lower = key_name.lower()
            # Try both upper and lower case from the config file
            if key_name_lower in config:
                return config[key_name_lower]
            elif key_name_upper in config:
                return config[key_name_upper]
            else:
                return None

    def get_openrouter_model(
        self,
        model_name,
        response_token_size=None,
        temperature=PRECISE_TEMPERATURE,
        streaming=False,
        callbacks=None,
    ):
        api_key = self.get_api_key("OPENROUTER_API_KEY")

        # Set the API base to the Open Router API, and set special headers that are required
        api_base = "https://openrouter.ai/api/v1"
        headers = {"HTTP-Referer": "https://aicodebot.dev", "X-Title": AICODEBOT_NO_EMOJI}

        return ChatOpenAI(
            openai_api_key=api_key,
            openai_api_base=api_base,
            model=model_name,
            max_tokens=response_token_size,
            temperature=temperature,
            streaming=streaming,
            callbacks=callbacks,
            # In order to get ConversationBufferMemory to work, we need to set the tiktoken model name
            tiktoken_model_name=self.tiktoken_model_name,
            model_kwargs={"headers": headers},
        )

    def get_model_token_limit(self, model_name):
        model_token_limits = {
            "openai/gpt-4": 8192,
            "openai/gpt-4-32k": 32768,
            "anthropic/claude-2": 100_000,
            "gpt-4": 8192,
            "gpt-4-32k": 32768,
            "gpt-3.5-turbo": 4096,
            "gpt-3.5-turbo-16k": 16384,
        }
        if model_name in model_token_limits:
            return model_token_limits[model_name]
        else:
            raise ValueError(f"Model {model_name} not found")

    def get_token_size(self, text):
        """Get the number of tokens in a string using the tiktoken library."""
        encoding = tiktoken.encoding_for_model(self.tiktoken_model_name)
        tokens = encoding.encode(text)
        return len(tokens)

    def read_model_config(self):
        # Figure out which model to use, based on the config file or environment variables
        config = read_config() or {}
        self.provider = os.getenv(
            "AICODEBOT_MODEL_PROVIDER", config.get("language_model_provider", self.DEFAULT_PROVIDER)
        )
        self.model_name = os.getenv("AICODEBOT_MODEL", config.get("language_model", self.DEFAULT_MODEL))

        # --------------------------- API key verification --------------------------- #
        if self.provider == self.OPENAI:
            key_name = "OPENAI_API_KEY"
        elif self.provider == self.OPENROUTER:
            key_name = "OPENROUTER_API_KEY"
        elif self.provider == self.HUGGINGFACE_HUB:
            key_name = "HUGGINGFACE_API_KEY"
        else:
            raise ValueError(f"Unrecognized provider: {self.provider}")

        if not self.get_api_key(key_name):
            raise ValueError(
                f"In order to use {self.provider}, you must set the {key_name} in your environment or config file"
            )

        # TODO: Verify the model

        return self.provider, self.model_name

    @property
    def tiktoken_model_name(self):
        if "/" in self.model_name:
            if self.model_name.startswith("openai/"):
                # For OpenAI models, this is as simple as stripping the prefix "openai/" from the model name
                return self.model_name.replace("openai/", "")
            else:
                # For non-OpenAI models, we set the model name to "gpt-4" for now. Seems to work.
                # Tested with anthropic/claude2
                return self.DEFAULT_MODEL
        else:
            return self.model_name

    def use_appropriate_sized_model(self, chain, token_size):
        current_model = self.model_name
        gpt_4_limit = self.get_model_token_limit("gpt-4") * 0.9
        gpt_4_32k_limit = self.get_model_token_limit("gpt-4-32k") * 0.9
        if current_model in ["gpt-4", "gpt-4-32k"]:
            if token_size > gpt_4_32k_limit:
                raise ValueError("Token limit exceeded for GPT4, try using less context (files)")
            elif token_size > gpt_4_limit:
                if "gpt-4-32k" in openai_supported_engines():
                    self.model_name = "gpt-4-32k"
                else:
                    raise ValueError(
                        "Your request is too large for gpt-4, and you don't have access to gpt-4-32k.\n"
                        "Hint: Try using openrouter.ai which has access to lots of models. See the README for details."
                    )
            else:
                self.model_name = "gpt-4"

        elif current_model in ["openai/gpt-4", "openai/gpt-4-32k"]:
            if token_size > gpt_4_32k_limit:
                raise ValueError(
                    "Token limit exceeded for GPT4, try using less context (files)\n"
                    "Hint: try anthropic/claude-2 (100k token limit)"
                )
            elif token_size > gpt_4_limit:
                self.model_name = "openai/gpt-4-32k"
            else:
                self.model_name = "openai/gpt-4"

        elif current_model in ["gpt-3.5-turbo", "gpt-3.5-turbo-16k"]:
            gpt_3_limit = self.get_model_token_limit("gpt-3.5-turbo") * 0.9
            gpt_3_16k_limit = self.get_model_token_limit("gpt-3.5-turbo-16k") * 0.9
            if token_size > gpt_3_16k_limit:
                raise ValueError("Token limit exceeded for GPT3.5, try using less context (files)")
            elif token_size > gpt_3_limit:
                self.model_name = "gp-3.5-turbo-16k"
            else:
                self.model_name = "gpt-3.5-turbo"

        if current_model != self.model_name:
            logger.trace(f"Switching from {current_model} to {self.model_name} to handle the context size.")
            chain.llm.model_name = self.model_name

        return current_model, self.model_name


def token_size(text):
    # Shortcut
    return LanguageModelManager().get_token_size(text)


# This is outside the class because functools.lru_cache doesn't work with class methods
# in Python <= 3.9
@functools.lru_cache(maxsize=1)
def openai_supported_engines():
    """Get a list of the models supported by the OpenAI API key."""
    config = read_config()
    openai.api_key = config["openai_api_key"]
    engines = engine.Engine.list()
    out = [engine.id for engine in engines.data]
    logger.trace(f"OpenAI supported engines: {out}")
    return out
